src/c/tb_client/context.zig

```diff
@@ -13,67 +30,256 @@ pub const ContextImplementation = struct {
 
-            const addresses = @ptrCast([*]const u8, addresses_ptr)[0..addresses_len];
-            context.thread.init(
-                allocator,
-                cluster_id,
-                addresses,
-                num_packets,
-                Context.on_completion,
-            ) catch |err| {
-                allocator.destroy(context);
-                return switch (err) {
-                    error.Unexpected => .unexpected,
-                    error.OutOfMemory => .out_of_memory,
-                    error.InvalidAddress => .invalid_address,
-                    error.SystemResources => .system_resources,
-                    error.NetworkSubsystemFailed => .network_subsystem,
+            log.debug("{}: init: initializing thread", .{context.client_id});
+            try context.thread.init(context);
+            errdefer context.thread.deinit(context.allocator);
+
+            return context;
+        }
+
+        pub fn deinit(self: *Context) void {
+            self.thread.deinit();
+
+            self.client.deinit(self.allocator);
+            self.message_pool.deinit(self.allocator);
+            self.io.deinit();
+
+            self.allocator.free(self.addresses);
+            self.allocator.free(self.packets);
+            self.allocator.destroy(self);
+        }
+
+        pub fn tick(self: *Context) void {
+            self.client.tick();
+        }
+
+        pub fn run(self: *Context) void {
+            while (!self.thread.signal.is_shutdown()) {
+                self.tick();
+                self.io.run_for_ns(config.tick_ms * std.time.ns_per_ms) catch |err| {
+                    log.err("{}: IO.run() failed: {s}", .{
+                        self.client_id,
+                        @errorName(err),
+                    });
+                    @panic("IO.run() failed");
                 };
+            }
+        }
+
+        pub fn request(self: *Context, packet: *Packet) void {
+            const message = self.message_pool.get_message();
+            defer self.message_pool.unref(message);
+            self.messages_available -= 1;
+
+            // Get the size of each request structure in the packet.data:
+            const event_size: usize = operation_event_size(packet.operation) orelse {
+                return self.on_complete(packet, error.InvalidOperation);
             };
 
-            var list = tb_packet_list_t{};
-            for (context.thread.packets) |*packet| {
-                list.push(tb_packet_list_t.from(packet));
+            // Make sure the packet.data size is correct:
+            const readable = @ptrCast([*]const u8, packet.data)[0..packet.data_size];
+            if (readable.len == 0 or readable.len % event_size != 0) {
+                return self.on_complete(packet, error.InvalidDataSize);
+            }
+
+            // Make sure the packet.data wouldn't overflow a message:
+            const writable = message.buffer[@sizeOf(Header)..][0..config.message_body_size_max];
```

Comment:
I added a new test validating the `tb_packet_status` results, and I found the client was able to send requests slightly larger than `config.message_body_size_max` because the message buffer includes an extra sector size  (defined at `message_pool.message_size_max_padded`).

I understand this extra size should not be used by the message's body, so I restricted the buffer's upper bound. as a fix. Is that correct?

Reply:
Yes, great catch! And its good to have those branches tested as well.